physical/cockroachdb/cockroachdb_ha.go

```diff
@@ -0,0 +1,197 @@
+package cockroachdb
+
+import (
+	"database/sql"
+	"fmt"
+	"sync"
+	"time"
+
+	"github.com/hashicorp/go-uuid"
+	"github.com/hashicorp/vault/sdk/physical"
+)
+
+const (
+	// The lock TTL matches the default that Consul API uses, 15 seconds.
+	// Used as part of SQL commands to set/extend lock expiry time relative to
+	// database clock.
+	CockroachDBLockTTLSeconds = 15
+
+	// The amount of time to wait between the lock renewals
+	CockroachDBLockRenewInterval = 5 * time.Second
+
+	// CockroachDBLockRetryInterval is the amount of time to wait
+	// if a lock fails before trying again.
+	CockroachDBLockRetryInterval = time.Second
+)
+
+// Verify backend satisfies the correct interfaces.
+var (
+	_ physical.HABackend = (*CockroachDBBackend)(nil)
+	_ physical.Lock      = (*CockroachDBLock)(nil)
+)
+
+type CockroachDBLock struct {
+	backend  *CockroachDBBackend
+	key      string
+	value    string
+	identity string
+	lock     sync.Mutex
+
+	renewTicker *time.Ticker
+
+	// ttlSeconds is how long a lock is valid for.
+	ttlSeconds int
+
+	// renewInterval is how much time to wait between lock renewals.  must be << ttl.
+	renewInterval time.Duration
+
+	// retryInterval is how much time to wait between attempts to grab the lock.
+	retryInterval time.Duration
+}
+
+func (c *CockroachDBBackend) HAEnabled() bool {
+	return c.haEnabled
+}
+
+func (c *CockroachDBBackend) LockWith(key, value string) (physical.Lock, error) {
+	identity, err := uuid.GenerateUUID()
+	if err != nil {
+		return nil, err
+	}
+	return &CockroachDBLock{
+		backend:       c,
+		key:           key,
+		value:         value,
+		identity:      identity,
+		ttlSeconds:    CockroachDBLockTTLSeconds,
+		renewInterval: CockroachDBLockRenewInterval,
+		retryInterval: CockroachDBLockRetryInterval,
+	}, nil
+}
+
+// Lock tries to acquire the lock by repeatedly trying to create a record in the
+// CockroachDB table. It will block until either the stop channel is closed or
+// the lock could be acquired successfully. The returned channel will be closed
+// once the lock in the CockroachDB table cannot be renewed, either due to an
+// error speaking to CockroachDB or because someone else has taken it.
+func (l *CockroachDBLock) Lock(stopCh <-chan struct{}) (<-chan struct{}, error) {
+	l.lock.Lock()
+	defer l.lock.Unlock()
+
+	var (
+		success = make(chan struct{})
+		errors  = make(chan error)
```

Comment:
You might want to make this channel buffered.  I saw some output from a static analysis tool recently showing that in the PG implementation, there's an issue here when tryToLock attempts to write to the error channel, but Lock has already returned due to stopCh being closed.  In addition, you could make that write to the error channel in Lock tolerate blockage by using the optional send syntax.

Reply:
I've taken your suggestions and made the channel buffered and the send non-blocking.